src/rules.jl

```diff
@@ -116,3 +123,55 @@ rrule(::Any, ::Vararg{Any}) = nothing
 # explicitly empty body to save the compiler that work.
 
 (::Core.kwftype(typeof(rrule)))(::Any, ::Any, ::Vararg{Any}) = nothing
+
+"""
+    RuleConfig{F, R, T}
+
+The configuration
+ - `F`: **frule-like**. This is singleton `typeof` a function which acts like `frule`, but
+   which functions via invoking an AD system. it must match the [`frule`](@ref) signature.
+   If you do not have such a function it must be set to `Nothing` instead.
+ - `R`: **rrule-like**. This is singleton `typeof` a function which acts like `rrule`, but
+   which functions via invoking an AD system. it must match the [`rrule`](@ref) signature.
+   If you do not have such a function it must be set to `Nothing` instead.
+ - `T`: **traits**. This should be a `Union` of all special traits needed for rules to be
+   allowed to be defined for your AD. If nothing special this should be set to `Union{}`.
+
+Rule authors can dispatch on this config when defining rules.
+For example:
+```julia
+# only define rrule for `pop!` on AD systems where mutation is supported.
+rrule(::RuleConfig{<:Any,<:Any,>:SupportsMutation}, typeof(pop!), ::Vector) = ...
+
+# this definition of map is for any AD that defines a forwards mode
+rrule(conf::RuleConfig{<:Function}, typeof(map), ::Vector) = ...
+
+# this definition of map is for any AD that only defines a reverse mode.
+# It is not as good as the rrule that can be used if the AD defines a forward-mode as well.
+rrule(conf::RuleConfig{Nothing,<:Function}, typeof(map), ::Vector) = ...
+```
+
+For more details see [rule configurations and calling back into AD](@ref config).
+"""
+abstract type RuleConfig{F<:Union{Function,Nothing}, R<:Union{Function,Nothing}, T} end
+
+# if no config is present then fallback to config-less rules
+frule(::RuleConfig, ārgs, f, args...; kwargs...) = frule(ārgs, f, args...; kwargs...))
+rrule(::RuleConfig, f, args...; kwargs...) = rrule(f, args...; kwargs...)
+
+function frule_via_ad(::RuleConfig{F}, ārgs, f, args...; kwargs...) where F <: Function
+    #TODO: Should this pass on the config? I suspect it should so we can use it for avoiding stack-overflows
```

Comment:
Would be good to construct a concrete example where a stack overflow occurs so that there's something to work around.

Reply:
yeah, this is WIP still haven't worked that one though.
That is why this is TODO